### PR TITLE
Separate Factors and Allow Negs

### PR DESCRIPTION
**Pull request by borcuttjahns**
@aarmey @andrewram4287 Sharing this now as a bookmark slash opportunity for you to play with the revised method. It works and nails the synthetic scene and IL-2 datasets but still seems to fail for the PopAlign stuff which I'll investigate. Let me know your thoughts and feel free to make any efficiency or clarity-related edits - just shoot me a message if you plan to!

### CODE REVIEW

**Comment by aarmey** [comment]
Excellent! Just a note that running Fig. 4 with anything other than rank 3 leads to an error. I think this may be an issue with `cluster_type()`, though, and not the changes in this PR.

**Comment by andrewram4287** [comment]
@aarmey I am currently running Fig4 with other ranks and I am not receiving an error on my end....

**Comment by aarmey** [comment]
Ok. Maybe I messed something up.

**Comment by borcuttjahns** [comment]
@aarmey @andrewram4287 Ok so I think this is currently working but there are still a couple of issues that should be talked about.

1) Because in cases where NK values vary by condition, I still have to make that condition factor positive by using an exponential. This sort of defeats the purpose of this development for that particular factor but negatives are still handled.
2) The method seems to work but despite the heightened freedom from separating means and covariances it seems to fit the PopAlign data with worse likelihood. It also (from mean factors) doesn't identify the "correct" divergent drugs although I haven't checked if it does from covariance factors.

I think I'm going to leave this here at least for the time being until we can meet to discuss - if either of you want to test it or improve it feel free!

**Comment by aarmey** [comment]
Regarding 1: I think our same reasoning applies and we should fully separate the nk factors from the rest.

Regarding 2: Now that we don't require positive values, it's likely better to use PCA instead of NMF for the RNAseq initial reduction. A nice additional benefit is it'll be way faster.

**Comment by aarmey** [comment]
Finally, I'm happy merging this if it looks like the flow analysis is improved. We don't really know what to expect with the RNAseq data, so I don't make much of surprising results there.

File: gmm/tests/test_GMM.py [hunk 1]
```diff
@@ -30,6 +30,40 @@
 )
 
 
+@pytest.mark.parametrize("rank", [3, 10])
+@pytest.mark.parametrize("nk_r", [True, False])
```
Comment (aarmey) [inline gmm/tests/test_GMM.py:34]: Can you move these down to their old position, so I can see the diff? Or, is there a reason these should be moved up?

File: gmm/figures/figure7.py [hunk 2]
```diff
@@ -48,9 +48,7 @@ def makeFigure():
     cluster_type(drugXA, fac, typeXA, ax=ax[2])
     norm_dist_plot(facXA["Dimension3"].to_pandas(), ax[10])
     reorder_table(facXA["Dimension3"].to_pandas(), ax[9])
-    labels = ["Cluster", "NMF", "Drug"]
-        
-    drug_gene_plot(facXA, "Budesonide", nfac, ax[5], max=True)
```
Comment (aarmey) [inline gmm/figures/figure7.py:53]: What was this and why did it get removed?

File: gmm/figures/figure4.py [hunk 3]
```diff
@@ -24,9 +24,9 @@ def makeFigure():
     flowXA, celltypeXA = smallDF(cellperexp)
     flowXA = flowXA.loc[marks,:,:,:,:]
     
-    rank = 3
-    n_cluster = 3
-
+    rank = 5
+    n_cluster = 6
+
```
Comment (aarmey) [inline gmm/figures/figure4.py:29]: Lint.

File: gmm/tensor.py [hunk 4]
```diff
@@ -1,14 +1,19 @@
+from re import S
 import numpy as np
 import pandas as pd
 import jax.numpy as jnp
 import jax.scipy.special as jsp
 import tensorly as tl
 from tqdm import tqdm
 import xarray as xa
+import jax.lax as jlx
```
Comment (aarmey) [inline gmm/tensor.py:9]: Are you using any of these new imports? They seem like very strange items...

**Review by aarmey — CHANGES_REQUESTED** [review]
Some questions.

File: gmm/figures/figure7.py (hunk 2 above)
Comment (andrewram4287) [reply gmm/figures/figure7.py:53]: No, I think it was on accident.

File: gmm/tensor.py (hunk 4 above)
Comment (andrewram4287) [reply gmm/tensor.py:9]: I removed a couple of imports!

File: gmm/tests/test_GMM.py (hunk 1 above)
Comment (andrewram4287) [reply gmm/tests/test_GMM.py:34]: I think @borcuttjahns moved them, but I can put them back to their original position if you'd like.

File: gmm/tensor.py [hunk 5]
```diff
@@ -9,6 +9,9 @@
 from jax import value_and_grad, jit
 from jax.lax.linalg import triangular_solve
 from scipy.optimize import minimize, Bounds
+from copy import copy
+from jax.experimental.host_callback import id_print
```
Comment (aarmey) [inline gmm/tensor.py:13]: I don't see `id_print` being used.

File: gmm/tensor.py [hunk 6]
```diff
@@ -72,10 +77,21 @@ def __init__(
             )
         )
 
+        vectorIn = vectorIn[nN[-1]::]
+        nN[2::] -= rank * shape[1]
+        nN = np.delete(nN, 2)
```
Comment (aarmey) [inline gmm/tensor.py:82]: I don't follow what's going on here. Why are you deleting the third element? Can you provide a comment for what this block does?

File: gmm/tensor.py [hunk 7]
```diff
@@ -72,10 +77,21 @@ def __init__(
...
+        covarFacS = list(shape)
+        covarFacS.pop(1)
```
Comment (aarmey) [inline gmm/tensor.py:84]: Why are you poping an element here?

File: gmm/tensor.py [hunk 8]
```diff
@@ -93,17 +109,17 @@ def get_covariances(self) -> jnp.ndarray:
...
-            self.factors[0],
+            self.covFacs[0],
             self.covars,
-            *self.factors[2::],)
+            *self.covFacs[1::],)
 
     def get_covariances_xarray(self, X):
         """Return covariance matrices."""
-        covar = jnp.einsum(
+        covar = np.einsum(
```
Comment (aarmey) [inline gmm/tensor.py:118]: This should almost certainly be `jnp`

File: gmm/tensor.py [hunk 9]
```diff
@@ -149,6 +165,7 @@ def sample(self, n_samples: int = 1000):
 
     def get_factors_xarray(self, X):
         cp_factors = tl.cp_normalize(self)
+        # cp_factors = self.reorient_factors(cp_factors)
```
Comment (aarmey) [inline gmm/tensor.py:168]: Remove unused code.

File: gmm/tensor.py [hunk 10]
```diff
@@ -165,6 +182,16 @@ def get_factors_xarray(self, X):
                         coords={"Cmp": cmpCol, **coordinates})
 
         return da
+    
+    def reorient_factors(self, X):
```
Comment (aarmey) [inline gmm/tensor.py:186]: This does not use anything from the class (no `self.X`). Should not be defined within the class.

Comment (aarmey) [reply gmm/tensor.py:186]: Also, [`cp_flip_sign`](https://github.com/tensorly/tensorly/blob/main/tensorly/cp_tensor.py#L329) does this for you.

**Review by aarmey — CHANGES_REQUESTED** [review]
More suggestions.

File: gmm/tensor.py (hunk 7 above)
Comment (andrewram4287) [reply gmm/tensor.py:84]: This is to get rid of the "signal" dimension. The covariance factor of (Signal x Signal x Rank) is rearranged in a different part of the code

File: gmm/tensor.py (hunk 6 above)
Comment (andrewram4287) [reply gmm/tensor.py:82]: Again, this is to ensure that we are not solving for the covariance of the signals.

**Comment by aarmey** [comment]
This looks close enough—merging!

**Pull request merged**